sdk/src/compile.rs

```diff
@@ -0,0 +1,180 @@
+            Self::Jolt => write!(f, "jolt"),
+        }
+    }
+}
+
+#[derive(Clone)]
+/// Options for dynamic compilation of guest programs.
+///
+/// By default, compilation occurs within `/tmp`. However, the implementation does respect the [`OUT_DIR`](https://doc.rust-lang.org/cargo/reference/environment-variables.html) environment variable.
+pub struct CompileOpts {
+    /// The (in-workspace) package to build.
+    pub package: String,
+    /// The binary produced by the build that should be loaded into the zkVM after successful compilation.
+    pub binary: String,
+    debug: bool,
+    native: bool,
+    unique: bool,
+    pub(crate) memlimit: Option<usize>, // in mb
+}
+
+impl CompileOpts {
+    /// Setup options for dynamic compilation.
+    pub fn new(package: &str, binary: &str) -> Self {
+        Self {
+            package: package.to_string(),
+            binary: binary.to_string(),
+            debug: false,
+            native: false,
+            unique: false,
+            memlimit: None,
+        }
+    }
+
+    /// Set dynamic compilation to build the guest program in a debug profile.
+    pub fn set_debug_build(&mut self, debug: bool) {
+        self.debug = debug;
+    }
+
+    /// Set dynamic compilation to build for the native (host machine) target, rather than for the zkVM.
+    pub fn set_native_build(&mut self, native: bool) {
+        self.native = native;
+    }
+
+    /// Set dynamic compilation to run a unique build that neither overwrites prior builds nor will be overwritten by future builds. May be used to concurrently build different versions of the same binary.
+    ///
+    /// Note: the SDK does not automatically clean or otherwise manage the resultant builds in the output directory.
+    pub fn set_unique_build(&mut self, unique: bool) {
+        self.unique = unique;
+    }
+
+    /// Set the amount of memory available to the guest program. For certain provers increasing the memory limit can lead to corresponding increases in the required proving time.
+    ///
+    /// Compilation will fail if this option is set when compiling for use with [`Jolt`](crate::jolt::Jolt), which uses a fixed memory size.
+    pub fn set_memlimit(&mut self, memlimit: usize) {
+        self.memlimit = Some(memlimit);
+    }
+
+    fn set_linker(&mut self, prover: &ForProver) -> Result<PathBuf, BuildError> {
+        let linker_script = match prover {
+            ForProver::Jolt => {
+                if self.memlimit.is_some() {
+                    return Err(BuildError::InvalidMemoryConfiguration);
+                }
+
+                include_str!("./linker-scripts/jolt.x").into()
+            }
+            ForProver::Default => {
+                if self.memlimit.is_none() {
+                    return Err(BuildError::InvalidMemoryConfiguration);
+                }
+
+                include_str!("./linker-scripts/default.x").replace(
+                    "{MEMORY_LIMIT}",
```

Comment:
I'm pretty sure it sets the size of the memory usable by the program, as well as the bytecode (since it's the same memory view). The assertion here is just scoped to capturing too large of a program in the original setup.